### PR TITLE
add map .keys .values .has_key list.remove, DCALL support

### PR DESCRIPTION
**Pull request by carltraveler**
a. add map.keys()
b. add map.has_key()
c. add map.values()
d. add list.remove(element)
e. add DCALL get the jump target from top stack

### CODE REVIEW

**Comment by Honglei-Cong** [comment]
For VM opcode changes, please make proposal on OEP firstly.

File: vm/neovm/func_array.go [hunk 1]
```diff
@@ -178,7 +185,133 @@ func opReverse(e *ExecutionEngine) (VMState, error) {
...
+		m := item.(*types.Map)
+		m.Remove(index)
+	case *types.Array:
+		m, err := item.GetArray()
+		if err != nil {
+			return FAULT, errors.NewErr("[opRemove]get Array error!")
+		}
+		atidx := -1
+		for i, obj := range m {
+			if index.Equals(obj) {
```
Comment (laizy) [inline vm/neovm/func_array.go]: seems really strange here.

File: vm/neovm/func_array.go [hunk 2]
```diff
@@ -178,7 +185,133 @@ func opReverse(e *ExecutionEngine) (VMState, error) {
...
+	case *types.Map:
+		mapitem, err := item.GetMap()
+		if err != nil {
+			return FAULT, err
+		}
+
+		var unsortKey []string
+		keyMap := make(map[string]types.StackItems, 0)
+		keys := make([]types.StackItems, len(mapitem))
+		for k := range mapitem {
```
Comment (laizy) [inline vm/neovm/func_array.go]: avoid duplicated code.

File: vm/neovm/func_flowcontrol.go [hunk 3]
```diff
@@ -68,6 +68,26 @@ func opCall(e *ExecutionEngine) (VMState, error) {
...
+	e.Context.SetInstructionPointer(int64(e.Context.GetInstructionPointer()))
+	e.PushContext(context)
+
+	dest, err := PopBigInt(e)
+	if err != nil {
+		return FAULT, errors.ERR_DCALL_OFFSET_ERROR
+	}
+	offset := dest.Int64()
+
+	if offset < 0 || int(offset) > len(e.Context.Code) {
```
Comment (laizy) [inline vm/neovm/func_flowcontrol.go]: is `>=`

File: vm/neovm/func_flowcontrol.go [hunk 4]
```diff
@@ -68,6 +68,26 @@ func opCall(e *ExecutionEngine) (VMState, error) {
...
+		return FAULT, errors.ERR_DCALL_OFFSET_ERROR
```
Comment (laizy) [inline vm/neovm/func_flowcontrol.go:82]: `offset` is confusing. it is an absolute position.

File: vm/neovm/func_validate.go [hunk 5]
```diff
@@ -467,6 +467,18 @@ func validatePickItem(e *ExecutionEngine) error {
...
 		}
+	case *types.ByteArray:
+		index, err := PeekBigInteger(e)
+		if err != nil {
+			return err
+		}
+		if index.Sign() < 0 {
+			return errors.ERR_BAD_VALUE
+		}
+		barr, err := item.GetByteArray()
```
Comment (laizy) [inline vm/neovm/func_validate.go:478]: use `barr, _ :=` or  check error

File: vm/neovm/func_validate.go [hunk 6]
```diff
@@ -629,3 +649,58 @@ func LogStackTrace(e *ExecutionEngine, needStackCount int, desc string) error {
...
+}
+
+func validateDCALL(e *ExecutionEngine) error {
+	dest, err := PeekBigInteger(e)
+	if err != nil {
+		return fmt.Errorf("validateDJMP error: %s", errors.ERR_NOT_SUPPORT_TYPE)
+	}
+	offset := dest.Int64()
+
+	if offset < 0 || int(offset) > len(e.Context.Code) {
```
Comment (laizy) [inline vm/neovm/func_validate.go]: `>=`

File: vm/neovm/func_array.go (hunk 2 above)
Comment (carltraveler) [reply vm/neovm/func_array.go]: ok, I will realize it with a function

File: vm/neovm/func_flowcontrol.go (hunk 3 above)
Comment (carltraveler) [reply vm/neovm/func_flowcontrol.go]: thanks. it is

File: vm/neovm/func_flowcontrol.go (hunk 4 above)
Comment (carltraveler) [reply vm/neovm/func_flowcontrol.go:82]: change it to target

File: vm/neovm/func_validate.go (hunk 6 above)
Comment (carltraveler) [reply vm/neovm/func_validate.go]: same obove

File: vm/neovm/func_flowcontrol.go (hunk 3 above)
Comment (carltraveler) [reply vm/neovm/func_flowcontrol.go]: oh, if offset = len(e.Context.code) will out of bound

File: vm/neovm/func_validate.go [hunk 7]
```diff
@@ -589,17 +604,25 @@ func validatorRemove(e *ExecutionEngine) error {
...
 
-	if _, ok := item.(*types.Map); !ok {
-		return errors.ERR_REMOVE_NOT_SUPPORT
+	switch item.(type) {
+	case *types.Map:
+		if !value.IsMapKey() {
+			return errors.ERR_NOT_MAP_KEY
+		}
+		if _, ok := item.(*types.Map); !ok {
+			return errors.ERR_REMOVE_NOT_SUPPORT
```
Comment (laizy) [inline vm/neovm/func_validate.go]: this is not needed. has guaranteed to be `*types.Map`.

File: vm/neovm/func_validate.go [hunk 8]
```diff
@@ -589,17 +604,25 @@ func validatorRemove(e *ExecutionEngine) error {
...
+		}
+	case *types.Array:
+		if _, ok := item.(*types.Array); !ok {
```
Comment (laizy) [inline vm/neovm/func_validate.go]: ditto

**Review by laizy — CHANGES_REQUESTED** [review]

**Review by tanZiWen — DISMISSED** [review]

**Comment by laizy** [comment]
add unit test?

**Comment by laizy** [comment]
missing `dcall` opcode test.

**Review by laizy — APPROVED** [review]

**Pull request merged**